fix polymer-2-application generated unit test

pghalliday opened the pull request:
Fixes #603 

 - [x] CHANGELOG.md has been updated

rictic commented on src/init/application/templates/polymer-2.x/test/_element/_element_test.html:
```diff
@@ -22,7 +22,7 @@
         test('instantiating the element works', function() {
           var element = fixture('basic');
-          assert.equal(element.is, '<%= elementName %>');
+          assert.equal(element.constructor.is, '<%= elementName %>');
```
I think there may be some issues with `constructor` when compiled to ES5?

How about something like:

```javascript
assert.true(!!customElements.get('<%= elementName %>'));
assert.true(element instanceof Polymer.Element);
```

rictic replied:
Hm, there's a few Polymer base classes. How about:

```javascript
assert.true(!!customElements.get('<%= elementName %>'));
assert.true(element instanceof customElements.get('<%= elementName %>'));
```

pghalliday replied:
This doesn't actually assert that the element is an `<%= elementName %>` does it?

Just that there is a such an element type and this element is a sub class  of `Polymer.Element`

Also i'm not sure that tests get compiled to ES5 (i've been trying to figure that out too) and the element under test is using ES6 classes

rictic replied:
My second suggestion addresses the concern that the element is an upgraded custom element of the expected tagname. We could add:

```javascript
assert.true(Polymer.telemetry.registrations.indexOf(element.localName) !== -1);
```

To also assert that the Polymer element class was finalized without issue (all the properties, observers, and templates have been associated to the class)

pghalliday replied:
I'll be honest - I have no clue what that line of code means :o

pghalliday replied:
Would this be better, for the sake of clarity and also in keeping with the intent of the original test:

```
assert.equal(element.localName, '<%= elementName %>');
```

That passes

pghalliday replied:
I would be concerned about putting a number of obtuse tests that might be hard to understand, given that this should really be an example for people (in reality they probably won't be testing the element name and will actually just be testing that it behaves as they expect)

rictic replied:
`element.localName` will pass pretty much no matter what, even if there's no registered custom element with that name. E.g. paste this into your devtools on this very page: `document.createElement('foo-bar').localName`

The Polymer telemetry assert is a little crazy, that's fair. What about just the other two?

```javascript
// Test that the element was registered properly.
assert.true(!!customElements.get('<%= elementName %>'));
assert.true(element instanceof customElements.get('<%= elementName %>'));
```

That'll end up looking like:

```javascript
// Test that the element was registered properly.
assert.true(!!customElements.get('my-element'));
assert.true(element instanceof customElements.get('my-element'));
```

pghalliday replied:
Thinking about it, a better example might be something like this

```
assert.equal(element.$.heading.textContent, 'Hello World!');
```

and adding the `prop1` to the fixture, and giving the `<h2>` in the element an id of `heading`

tony19 replied:
Instead of `assert.true(!!foo)`, how about [`assert.isOk(foo)`](http://chaijs.com/api/assert/#isokobject-message)?

rictic commented:
Thanks for the PR!

Would you mind also updating the element template too?

pghalliday commented:
@rictic, what's wrong with the element template?

rictic commented:
It looks like it has the same test assertion as the application template

pghalliday commented:
ah, I think iwas confused about what the element template was :) I'll take a look

rictic commented:
It's a sibling directory to `application`

pghalliday commented:
How about this new test, I think it's a better example of a good unit test as it tests the behavior

rictic commented:
I think an assert with `customElements` is better as a first test, as it gives very useful diagnostic info when it fails. i.e. that registering the element failed.

There's lots of reasons why some text content might not be right.

I wouldn't be opposed to adding the text content test as an additional test.

pghalliday commented:
Interestingly, when I `console.log(element.textContent)` I don't get anything that I can really use so I don't know what I should expect there. The reason I added the `id` attribute to the `h2` was to provide some abstraction from the internal implementation and make it part of the public interface - but I see your point.

Anyway I have reverted that and just fixed the instantiation test.

FredKSchott commented:
Thanks for the contribution @pghalliday! Closing this issue because this was just fixed in a greater rewrite of the application templates to make their tests more thorough: https://github.com/Polymer/polymer-cli/pull/649
